asmcomp/debug/available_regs.ml

```diff
@@ -0,0 +1,387 @@
+   unreachable. Having a sum type to represent that may be cleaner.
+   I don't think this allocation cost would really matter *)
+
+let augment_availability_at_raise avail =
+  avail_at_raise := RAS.inter avail !avail_at_raise
+
+(* [available_regs ~instr ~avail_before] calculates, given the registers
+   "available before" an instruction [instr], the registers that are available
+   after [instr].  This is a forwards dataflow analysis.
+
+   "available before" can be thought of, at the assembly level, as the set of
+   registers available when the program counter is equal to the address of the
+   particular instruction under consideration (that is to say, immediately
+   prior to the instruction being executed).  Inputs to that instruction are
+   available at this point even if the instruction will clobber them.  Results
+   from the previous instruction are also available at this point.
+
+   The [available_before] field of each instruction is updated by this
+   function.
+*)
+let rec available_regs (instr : M.instruction)
+      ~(avail_before : RAS.t) : RAS.t =
+  let avail_before : RAS.t =
+    (* If this instruction might expand into multiple machine instructions
+       and clobber registers during that code sequence, make sure any
+       such clobbered registers are marked as "unavailable before". *)
+    (* CR-someday mshinwell: We could maybe try to improve on this in the
+       future, so e.g. when stopped immediately before an allocation,
+       registers aren't marked unavailable when not yet clobbered by the
+       allocation code.
+       Should we be using "available across" instead of "available before"?
+       It's not clear that actually solves the problem. *)
+    (* XCR pchambart: This test should probably surround everything,
+       especialy in the case where it is changed to a sum type. *)
+    match avail_before with
+    | Unreachable -> Unreachable
+    | Ok avail_before ->
+      (* Every register that is live across an instruction should also be
+         available before the instruction.  (We cannot assert that a register
+         should not be an input to an instruction unless it is available due
+         to the special case below.) *)
+      (* XCR pchambart: The difference between live and avail_before is
+         probably due to the 'interesting' filter. *)
+      if not (R.Set.subset instr.live (RD.Set.forget_debug_info avail_before))
+      then begin
+        Misc.fatal_errorf "Live registers not a subset of available registers: \
+            live={%a} avail_before=%a missing={%a} insn=%a"
+          Printmach.regset instr.live
+          (RAS.print ~print_reg:Printmach.reg)
+          (RAS.Ok avail_before)
+          Printmach.regset (R.Set.diff instr.live
+            (RD.Set.forget_debug_info avail_before))
+          Printmach.instr ({ instr with M. next = M.end_instr (); })
+      end;
+      match instr.desc with
+      | Iop (Ialloc _) ->
```

Comment:
Across an allocation, only live and spilled values are guaranteed to be available.

Reply:
This should be fixed now